tsdb/chunks/head_chunks_test.go

```diff
@@ -0,0 +1,301 @@
+		_, err = chkCRC32.Write(buf[:MintMaxtSize])
+		testutil.Ok(t, err)
+
+		expectedBytes = append(expectedBytes, byte(chunk.Encoding()))
+		_, err = chkCRC32.Write([]byte{byte(chunk.Encoding())})
+		testutil.Ok(t, err)
+
+		n := binary.PutUvarint(buf[:], uint64(len(chunk.Bytes())))
+		expectedBytes = append(expectedBytes, buf[:n]...)
+		_, err = chkCRC32.Write(buf[:n])
+		testutil.Ok(t, err)
+
+		expectedBytes = append(expectedBytes, chunk.Bytes()...)
+		_, err = chkCRC32.Write(chunk.Bytes())
+		testutil.Ok(t, err)
+
+		expectedBytes = append(expectedBytes, chkCRC32.Sum(nil)...)
+
+		// += encoding + chunk data len + chunk data + CRC + seriesRef,mint,maxt of next chunk.
+		nextChunkOffset += 1 + uint64(n) + uint64(len(chunk.Bytes())) + CRCSize + SeriesRefSize + 2*MintMaxtSize
+	}
+	testutil.Ok(t, hrw.flushBuffer())
+
+	testutil.Assert(t, len(hrw.mmappedChunkFiles) == 1 && len(hrw.closers) == 1, "expected only 1 mmapped file, got %d", len(hrw.mmappedChunkFiles))
+
+	fileName := hrw.curFile.Name()
+	actualBytes, err := ioutil.ReadFile(fileName)
+	testutil.Ok(t, err)
+
+	// Check header of the segment file.
+	testutil.Equals(t, MagicHeadChunks, int(binary.BigEndian.Uint32(actualBytes[0:MagicChunksSize])))
+	testutil.Equals(t, chunksFormatV1, int(actualBytes[MagicChunksSize]))
+
+	mint := binary.BigEndian.Uint64(actualBytes[HeaderMintOffset : HeaderMintOffset+8])
+	maxt := binary.BigEndian.Uint64(actualBytes[HeaderMaxtOffset : HeaderMaxtOffset+8])
+	testutil.Assert(t, mint > 0, "expected mint >0")
+	testutil.Assert(t, maxt == 0, "expected maxt = 0 for an active file")
+
+	// Remaining chunk data.
+	testutil.Equals(t, expectedBytes, actualBytes[HeadChunkFileHeaderSize:])
+}
+
+func TestHeadReadWriter_ReadChunk(t *testing.T) {
+	hrw, close := testHeadReadWriter(t)
+	defer func() {
+		testutil.Ok(t, hrw.Close())
+		close()
+	}()
+
+	type expectedDataType struct {
+		chunkRef uint64
+		chunk    chunkenc.Chunk
+	}
+
+	expectedData := []expectedDataType{}
+	numChunks := 100
+
+	timesRan := 0
+	populateChunks := func() {
+		for i := 0; i < numChunks; i++ {
+			chunk := randomChunk(t)
+			chunkRef, err := hrw.WriteChunk(
+				uint64(rand.Int63()),
+				int64(((timesRan*numChunks)+i)*1000+1),
+				int64(((timesRan*numChunks)+i+1)*1000),
+				chunk,
+			)
+			testutil.Ok(t, err)
+
+			expectedData = append(expectedData, expectedDataType{
+				chunkRef: chunkRef,
+				chunk:    chunk,
+			})
+		}
+	}
+
+	// Add chunks till we fill at least 1 segment file
+	// and also have something in the in-memory buffer.
+	for hrw.curFileSequence < 2 || hrw.wbuf.Buffered() == 0 {
+		populateChunks()
+		timesRan++
+	}
+
+	// This is to test the access of buffered chunks.
+	testutil.Assert(t, hrw.wbuf.Buffered() > 0, "there are no buffered chunks")
+
+	for _, exp := range expectedData {
+		actChunk, err := hrw.Chunk(exp.chunkRef)
+		testutil.Ok(t, err)
+		testutil.Equals(t, exp.chunk.Bytes(), actChunk.Bytes())
+	}
+}
+
+func TestHeadReadWriter_IterateChunks(t *testing.T) {
```

Comment:
the same here you jast add an additional check for the iterator. the boostrap is quite similiar.

If you prefer to keep those separate can't you make the data boostrapper in a seprate func to be used in all tests?

Reply:
Yeah, so this can be combined with `ReadChunks` test and would work out well too as in the read test we are assuring >1 files. I will combine these 2.